lib/awful/widget/common.lua

```diff
@@ -117,11 +118,40 @@ function common.list_update(w, buttons, label, data, objects, args)
         local text, bg, bg_image, icon, item_args = label(o, cache.tb)
         item_args = item_args or {}
 
-        -- The text might be invalid, so use pcall.
         if cache.tbm and (text == nil or text == "") then
-            cache.tbm:set_margins(0)
-        elseif cache.tb then
-            if not cache.tb:set_markup_silently(text) then
+            -- Remove text margin when there is no text
+
+            local current_margins = { left = cache.tbm:get_left(),
+                                      right = cache.tbm:get_right(),
+                                      top = cache.tbm:get_top(),
+                                      bottom = cache.tbm:get_bottom() }
+
+            local has_any_margin = false
+            for _, m in pairs(current_margins) do
+                if m > 0 then
+                    has_any_margin = true
+                    break
+                end
+            end
+
+            if has_any_margin then
+                -- Save current margin values so they can be restored in case
+                -- text will be present in future update
+                cache.tbm.saved_margins = current_margins
```

Comment:
Btw, I don't have much experience with lua, so:
Is it OK to add custom fields to existing objects like that?